RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc

```diff
@@ -0,0 +1,307 @@
+			fixedErrorRescalingForHits_(iConfig.getParameter<double>("fixedErrorRescaleFactorForHits")),
+			fixedErrorRescalingForHitless_(iConfig.getParameter<double>("fixedErrorRescaleFactorForHitless")),
+			adjustErrorsDyanmicallyForHits_(iConfig.getParameter<bool>("adjustErrorsDyanmicallyForHits")),
+			adjustErrorsDyanmicallyForHitless_(iConfig.getParameter<bool>("adjustErrorsDyanmicallyForHitless")),
+		    estimatorName_(iConfig.getParameter<std::string>("estimator")),
+			measurementTrackerTag_(consumes<MeasurementTrackerEvent>(iConfig.getParameter<edm::InputTag>("MeasurementTrackerEvent"))),
+			minEtaForTEC_(iConfig.getParameter<double>("minEtaForTEC")),
+			maxEtaForTOB_(iConfig.getParameter<double>("maxEtaForTOB")),
+			useHitlessSeeds_(iConfig.getParameter<bool>("UseHitlessSeeds")),
+			dummyPlane_(Plane::build(Plane::PositionType(), Plane::RotationType())){
+	produces<std::vector<TrajectorySeed> >();
+	updator_ = new KFUpdator();
+	foundCompatibleDet_=false;
+	numSeedsMade_=0;
+	theCategory = "Muon|RecoMuon|TSGForOI";
+}
+
+
+TSGForOI::~TSGForOI(){
+	if(updator_) delete updator_;
+}
+
+
+void TSGForOI::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+	iSetup.get<IdealMagneticFieldRecord>().get(magfield_);
+	iSetup.get<TrackingComponentsRecord>().get("PropagatorWithMaterial", propagatorOpposite_);
+	iSetup.get<TrackingComponentsRecord>().get("PropagatorWithMaterial", propagatorAlong_);
+	iSetup.get<GlobalTrackingGeometryRecord>().get(geometry_);
+    iSetup.get<TrackingComponentsRecord>().get(estimatorName_,estimator_);
+	iEvent.getByToken(measurementTrackerTag_, measurementTracker_);
+	edm::Handle<reco::TrackCollection> l2TrackCol;					iEvent.getByToken(src_, l2TrackCol);
+
+//	The product:
+	std::auto_ptr<std::vector<TrajectorySeed> > result(new std::vector<TrajectorySeed>());
+
+//	Get vector of Detector layers once:
+	std::vector<BarrelDetLayer const*> const& tob = measurementTracker_->geometricSearchTracker()->tobLayers();
+	std::vector<ForwardDetLayer const*> const& tecPositive = measurementTracker_->geometricSearchTracker()->posTecLayers();
+	std::vector<ForwardDetLayer const*> const& tecNegative = measurementTracker_->geometricSearchTracker()->negTecLayers();
+
+//	Get the suitable propagators:
+	std::unique_ptr<Propagator> propagatorAlong = SetPropagationDirection(*propagatorAlong_,alongMomentum);
+	std::unique_ptr<Propagator> propagatorOpposite = SetPropagationDirection(*propagatorOpposite_,oppositeToMomentum);
+
+//	Loop over the L2's and making seeds for all of them:
+//	edm::LogDebug(theCategory) << "TSGForOI::produce: Number of L2's: " << l2TrackCol->size();
+	for (unsigned int l2TrackColIndex(0);l2TrackColIndex!=l2TrackCol->size();++l2TrackColIndex){
+		const reco::TrackRef l2(l2TrackCol, l2TrackColIndex);
+		std::auto_ptr<std::vector<TrajectorySeed> > out(new std::vector<TrajectorySeed>());
+
+		FreeTrajectoryState fts = trajectoryStateTransform::initialFreeState(*l2, magfield_.product());
```

Comment:
both fts and notUpdatedFts start from the same call ```trajectoryStateTransform::innerFreeState(*l2,magfield_.product());```
I suggested to make a copy above and use the copy instead of making a complete computation